### PR TITLE
Standalone Lite: DDP Strategy Family 

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds the DDP family of strategies supported in Lite. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/strategies/utils.py [hunk 1]
```diff
@@ -31,12 +31,12 @@ def on_colab_kaggle() -> bool:
...
-        if precision == PrecisionType.BFLOAT:
-            return tensor.bfloat16()
+def _call_register_strategies(registry: _StrategyRegistry, base_module: str) -> None:
+    # TODO(lite): Remove this function once PL strategies inherit from Lite's Strategy base class
+    module = importlib.import_module(base_module)
+    for _, mod in getmembers(module, isclass):
+        if issubclass(mod, Strategy) and _is_register_method_overridden(mod, Strategy, "register_strategies"):
+            mod.register_strategies(registry)
 
     return tensor
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/utils.py:41]: Did you forget to remove this return?

```suggestion

```

File: src/lightning_lite/strategies/utils.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Copyright The PyTorch Lightning team.
```
Comment (carmocca) [inline src/lightning_lite/strategies/utils.py:1]: This file is in the strategies directory, but wouldn't a precision dir be more appropriate?

File: src/lightning_lite/strategies/fairscale.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        strategy_registry.register(
+            "ddp_sharded_find_unused_parameters_false",
+            cls,
+            description="DDP Sharded Strategy with `find_unused_parameters` as False",
+            find_unused_parameters=False,
+        )
+        strategy_registry.register(
+            "ddp_sharded",
+            cls,
+            description=f"{cls.__class__.__name__}",
```
Comment (carmocca) [inline src/lightning_lite/strategies/fairscale.py:109]:
```suggestion
            description=cls.__class__.__name__",
```

File: src/lightning_lite/strategies/fairscale.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+from torch.optim import Optimizer
+
+from lightning_lite.accelerators import Accelerator
+from lightning_lite.plugins import CheckpointIO, ClusterEnvironment, Precision
+from lightning_lite.strategies.ddp import DDPStrategy
+from lightning_lite.utilities.enums import PrecisionType
+from lightning_lite.utilities.imports import _IS_WINDOWS
+
+_FAIRSCALE_AVAILABLE = not _IS_WINDOWS and module_available("fairscale.nn")
+_FAIRSCALE_OSS_FP16_BROADCAST_AVAILABLE = _FAIRSCALE_AVAILABLE and compare_version("fairscale", operator.ge, "0.3.3")
```
Comment (carmocca) [inline src/lightning_lite/strategies/fairscale.py:32]: Our min supported fairscale is higher than 0.3.3

```suggestion
```

I removed this in #14476 so must be a bad merge

File: src/lightning_lite/strategies/ddp.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        strategy_registry.register(
+            "ddp_find_unused_parameters_false",
+            cls,
+            description="DDP Strategy with `find_unused_parameters` as False",
+            find_unused_parameters=False,
+        )
+        strategy_registry.register(
+            "ddp",
+            cls,
+            description=f"{cls.__class__.__name__}",
```
Comment (carmocca) [inline src/lightning_lite/strategies/ddp.py:160]:
```suggestion
            description=cls.__class__.__name__",
```

**Review by carmocca — APPROVED** [review]
Lots of code moving so might have missed something!

File: src/lightning_lite/strategies/ddp.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        rank_zero_only.rank = self.global_rank
+
+        self._process_group_backend = self._get_process_group_backend()
+        assert self.cluster_environment is not None
+        init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+
+    def _get_process_group_backend(self) -> str:
+        return (
+            self._process_group_backend
+            or _get_process_group_backend_from_env()
```
Comment (carmocca) [inline src/lightning_lite/strategies/ddp.py:179]: This should be removed from here, as it'll be removed in PL too: https://github.com/Lightning-AI/lightning/pull/14693

File: src/lightning_lite/strategies/fairscale.py (hunk 4 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/fairscale.py:32]: Okay, even better

File: src/pytorch_lightning/strategies/utils.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/strategies/utils.py:41]: ah, this came from a merge conflict

File: src/lightning_lite/strategies/fairscale.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/fairscale.py:109]: In all of your suggestions of past PRs, you have tricked me by keeping the last double quote. This time, I won't fall for it!

Comment (carmocca) [reply src/lightning_lite/strategies/fairscale.py]: lol! my bad, didn't notice it

File: src/lightning_lite/strategies/fairscale.py (hunk 4 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/fairscale.py]: Ok, checked the PR. Nothing else was missed, only these imports in the strategy here directly. updated

File: src/lightning_lite/strategies/utils.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/utils.py]: yes, why not

File: src/lightning_lite/plugins/precision/utils.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+
+from lightning_lite.utilities.enums import PrecisionType
+
+
+def _fp_to_half(tensor: torch.Tensor, precision: PrecisionType) -> torch.Tensor:
```
Comment (Borda) [inline src/lightning_lite/plugins/precision/utils.py:20]:
```suggestion
def _float_to_half(tensor: torch.Tensor, precision: PrecisionType) -> torch.Tensor:
```

File: src/lightning_lite/strategies/deepspeed.py [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+        .. warning:: ``DeepSpeedStrategy`` is in beta and subject to change.
+
+        Defaults have been set to enable ZeRO-Offload and some have been taken from the link below.
+        These defaults have been set generally, but may require tuning for optimum performance based on your model size.
+        `For more information: https://www.deepspeed.ai/docs/config-json/#zero-optimizations-for-fp16-training`.
+
+        Arguments:
+
+            zero_optimization: Enable ZeRO optimization. This is compatible with either `precision=16` or
+                `precision="bf16"`.
```
Comment (Borda) [inline src/lightning_lite/strategies/deepspeed.py:110]:
```suggestion
            zero_optimization: Enable ZeRO optimization. This is compatible with either ``precision=16`` or
                ``precision="bf16"``.
```

**Review by Borda — APPROVED** [review]

File: src/lightning_lite/plugins/precision/utils.py (hunk 7 above)
Comment (carmocca) [reply src/lightning_lite/plugins/precision/utils.py:20]: float is misleading, as there's many float types but it's usually meant for 32 bit. fp or floating_point is more generic: https://pytorch.org/docs/stable/generated/torch.is_floating_point.html
